src/Microsoft.DotNet.Interactive.Jupyter/Http/JupyterHttpConnection.cs

```diff
@@ -0,0 +1,164 @@
+        this(new HttpApiClient(serverUri, authProvider), authProvider)
+    { }
+
+    public JupyterHttpConnection(HttpApiClient apiClient, IAuthorizationProvider authProvider)
+    {
+        _apiClient = apiClient ?? throw new ArgumentNullException(nameof(apiClient));
+        _authProvider = authProvider ?? throw new ArgumentNullException(nameof(authProvider));
+
+        _disposables = new CompositeDisposable
+        {
+            _apiClient
+        };
+    }
+
+    public void Dispose()
+    {
+        Task.Run(ShutdownJupyterSessionsAsync).Wait();
+        _disposables.Dispose();
+    }
+
+    public async Task<IEnumerable<KernelSpec>> GetKernelSpecsAsync()
+    {
+        if (_availableKernels == null)
+        {
+            HttpResponseMessage response = await _apiClient.SendRequestAsync(
+                relativeApiPath: "api/kernelspecs",
+                content: null,
+                method: HttpMethod.Get
+            );
+
+            if (!response.IsSuccessStatusCode)
+            {
+                return Array.Empty<KernelSpec>();
+            }
+
+            byte[] bytes = await response.Content.ReadAsByteArrayAsync();
+            var list = JsonSerializer.Deserialize<KernelSpecs>(bytes);
+            _availableKernels = list?.kernelspecs?.Select(t => {
+                var spec = t.Value.spec;
+                spec.Name ??= t.Key;
+                return spec;
+            });
+        }
+
+        return _availableKernels;
+    }
+
+    public async Task<IJupyterKernelConnection> CreateKernelConnectionAsync(string kernelSpecName)
+    {
+        string kernelId = Guid.NewGuid().ToString();
+        var body = new
+        {
+            kernel = new
+            {
+                id = kernelId,
+                name = kernelSpecName
+            },
+            name = "",
+            path = $"dotnet-{kernelId}",
+            type = "notebook"
+        };
+
+        try
+        {
+            HttpResponseMessage response = await _apiClient.SendRequestAsync(
+                relativeApiPath: "api/sessions",
+                content: new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json"),
+                method: HttpMethod.Post
+            );
+
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new KernelStartException(kernelSpecName, response.ReasonPhrase);
+            }
+
+            byte[] bytes = await response.Content.ReadAsByteArrayAsync();
+            var session = JsonSerializer.Deserialize<KernelSessionInfo>(bytes);
+            _activeSessions.Add(session.id);
+            var kernelApiClient = _apiClient.CreateClient($"api/kernels/{session.kernel.id}");
+
+            return new JupyterKernelHttpConnection(kernelApiClient, _authProvider);
+        }
+        catch (Exception e)
+        {
+            throw new KernelStartException(kernelSpecName, e.Message);
```

Comment:
I don't think we need to wrap this in a custom exception. Can we just drop the whole `try`/`catch` or are we handling `KernelStartException` somewhere?

Reply:
Makes sense. The context for kernelSpecName is already in the command being executed. Will remove this in a separate PR.